### PR TITLE
sqlite: Add rollback support

### PR DESCRIPTION
**Pull request by brunnre8**
This enabled db migrations to be undone, or "down migrated". The down migration shouldn't be done automatically as it could lead to severe data loss if that were done. Hence, we still hard fail if we encounter a version lower than what we have in the DB.

A CLI will be added in a later commit that allows users to explicitly do that.

### CODE REVIEW

**Comment by brunnre8** [comment]
@itsjohncs well... if we want to have down migrations, that's what it takes I guess.
Bunch of code, CLI is still missing.

I still don't think that auto migration in the downward direction makes any sense, so we hard fail still.

A later PR can add a CLI that calls `storage.downgrade()`, the infra is in place.

Happy with that approach?

File: server/plugins/messageStorage/sqlite.ts [hunk 1]
```diff
@@ -198,6 +251,116 @@ class SqliteMessageStorage implements SearchableMessageStorage {
...
+			join migrations on migrations.id=rollback_steps.migration_id
+			where version > ?
+			order by version desc, step asc`,
+			since_version
+		);
+		const result: Rollback[] = [];
+
+		// convert to Rollback[]
+		// requires ordering in the sql statement
+		res.reduce((combined: Rollback[], cur) => {
```
Comment (itsjohncs) [inline server/plugins/messageStorage/sqlite.ts:267]: Is this reduction just being used as a for loop? This is pretty hard for me to follow.

File: server/plugins/messageStorage/sqlite.ts [hunk 2]
```diff
@@ -198,6 +251,116 @@ class SqliteMessageStorage implements SearchableMessageStorage {
...
+			await this.serialize_run("ROLLBACK", []);
+			throw err;
+		}
+
+		await this.serialize_run("COMMIT", []);
+		return new_version;
+	}
+
+	async downgrade() {
+		return this.downgrade_to(currentSchemaVersion);
```
Comment (itsjohncs) [inline server/plugins/messageStorage/sqlite.ts:335]: I think this and `return await this.downgrade_to(currentSchemaVersion)` is equivalent thanks to the fact that returning a promise from an async function implicitly chains that promise with the parent function's promise. But it strikes me as one of the more esoteric facts about JS and I'd think making `downgrade` a non-async function or awaiting `downgrade_to`'s promise would be clearer.

**Review by itsjohncs — APPROVED** [review]
Looks good. Happy to punt the work of deeper testing and making the CLI for when it's actually needed.

File: server/plugins/messageStorage/sqlite.ts (hunk 1 above)
Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts:267]: Yes.
There I was, being pleased with myself that I'm a real JS programmer that uses all the functional thingies....

okayokay, I'm gonna rewrite it with a plain boring loop then :wink:

File: server/plugins/messageStorage/sqlite.ts (hunk 2 above)
Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts:335]: ok, I'll await the result then... don't really want to switch to promises if I don't have to

File: server/plugins/messageStorage/sqlite.ts (hunk 1 above)
Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts]: done

**Pull request merged**